cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/OperationRuntime.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright © 2023 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.operation;
+
+
+import com.google.inject.Inject;
+import io.cdap.cdap.proto.id.OperationRunId;
+import io.cdap.cdap.proto.operation.OperationError;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.annotation.Nullable;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provides management of running operations. Stores {@link OperationController} in memory.
+ */
+public class OperationRuntime {
+
+  private final ReadWriteLock runtimeLock;
+  private final Map<OperationRunId, OperationController> controllers;
+  private final OperationRunner runner;
+  private final OperationStatePublisher statePublisher;
+
+  private static final Logger LOG = LoggerFactory.getLogger(OperationRuntime.class);
+
+  @Inject
+  OperationRuntime(OperationRunner runner, OperationStatePublisher statePublisher) {
+    this.runner = runner;
+    this.statePublisher = statePublisher;
+    this.runtimeLock = new ReentrantReadWriteLock();
+    this.controllers = new HashMap<>();
+  }
+
+  /** Runs an operation given the detail.
+   *
+   * @param runDetail detail of the run
+   * @return {@link OperationController} for the run
+   */
+  public OperationController run(OperationRunDetail runDetail) {
+    OperationController controller = getController(runDetail);
+    if (controller != null) {
+      LOG.debug("Operation has already been started: {}", runDetail.getRunId());
+      return controller;
+    }
+    try {
+      updateController(runDetail.getRunId(), runner.run(runDetail));
+    } catch (IllegalStateException e) {
+      statePublisher.publishFailed(runDetail.getRunId(),
+          new OperationError(e.getMessage(), Collections.emptyList())
+      );
+    }
+    return controllers.get(runDetail.getRunId());
+  }
+
+  /**
+   * Get controller for a given {@link OperationRunId}.
+   */
+  @Nullable
+  public OperationController getController(OperationRunDetail detail) {
+    Lock lock = this.runtimeLock.readLock();
+    lock.lock();
+    try {
+      return controllers.get(detail.getRunId());
```

Comment:
after a restart, this map will be empty. If we want to be able to control an operation after a restart, we will need to be able to create a controller for a run at any point.

Also, this looks like a good place for a LoadingCache. Don't need all the manual locking.

Reply:
I have changed the `OperationRunTimeService` to be an interface. In case of In-Memory implementation we cannot control an operation after a restart, rather the service startup will handle them. For worker based implementation this will be handeled